test/test_recipes.py

```diff
@@ -0,0 +1,331 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import OrderedDict
+from typing import Dict, Tuple
+
+import numpy as np
+import pytest
+
+import funsor.ops as ops
+from funsor.domains import Bint, Real, Reals
+from funsor.montecarlo import extract_samples
+from funsor.recipes import forward_filter_backward_rsample
+from funsor.terms import Lambda, Variable
+from funsor.testing import assert_close, random_gaussian
+from funsor.util import get_backend
+
+
+def get_moments(samples):
+    reduced_vars = frozenset(["particle"])
+    moments = OrderedDict()
+
+    # Compute first moments.
+    diffs = OrderedDict()
+    for name, value in samples.items():
+        mean = value.reduce(ops.mean, reduced_vars)
+        moments[name] = mean
+        diffs[name] = value - mean
+
+    # Compute centered second moments.
+    for i, (name1, diff1) in enumerate(diffs.items()):
+        diff1_ = diff1.reshape((diff1.output.num_elements, 1))
+        for name2, diff2 in list(diffs.items())[:i]:
+            diff_2 = diff2.reshape((1, diff2.output.num_elements))
+            diff12 = diff1_ * diff_2
+            moments[name1, name2] = diff12.reduce(ops.mean, reduced_vars)
+
+    return moments
+
+
+def check_ffbr(factors, eliminate, plates, actual_samples, actual_log_prob):
+    """
+    This can be seen as performing naive tensor variable elimination by
+    breaking all plates and creating a single flat joint distribution.
```

Comment:
If this actually turns out to be true, at some point we should move this logic into `funsor.sum_product` and test it against existing implementations since it's somewhat complex for testing logic and it is used to check correctness of what will presumably end up as one of our most important code paths if/when `AutoGaussian` is finished.

Reply:
SGTM, it seems reasonable to eventually factor this out as a `naive_sum_product` or sth. Before we refactor, we should fix the intractable tests below, as that will involve more lambdas etc.